### ISSUE TITLE
can we add option to set environment variable for mux domains, specially ssh domains, e.g. SSH_AUTH_SOCK

### ISSUE THREAD

**Issue by nerditation** [wez/wezterm#3332]
**Is your feature request related to a problem? Please describe.**

yes. on Windows, there are different kinds of ssh agents, and they all use different protocols. for instance, pageant uses window messages (good old Win32), ssh-agent (from the microsoft distributed openssh) uses named pipe (default to `\\.\pipe\openssh-ssh-agent`), wsl1 uses AF_UNIX socket, and wsl2 is essentially a VM on top of Windows' hypervisor and need some kind of forwarder.

there are some programs (e.g. I am using KeeAgent for KeePass) which speaks multiple agent protocols at the same time, but you have to be careful about your `SSH_AUTH_SOCK` environment variable. if you just work in a single environment, this should not be a problem, e.g. I can unset ·SSH_AUTH_SOCK` on Windows and let openssh use the default pipe, yet set the variable to the correct path in the shell startup script (e.g. ~/.bash_profile) in wsl1, etc.

but wezterm a terminal emulator with mux and ssh support, the builtin ssh implementation uses AF_UNIX socket:  when I use the local domain, I typically use the system openssh (both for remote shell and for git), yet when I connect to a remote session, I need to tell wezterm to use the unix socket (the same socket for wsl1 environment).

**Describe the solution you'd like**

currently, we can configure wezterm to set environment variables at startup, and for launch menu entries (`SpawnCommand`), but not for `SshDomain`. I would suggest add the same `set_environment_variables` field to `SshDomain`. or maybe we could also add options like `unset_environment_variables` too, like what #3222 suggested. as for other mux domain types (unix, tls), I don't know if such options could be generally useful or not.

**Describe alternatives you've considered**

we could add a specific configuration like `SshDomain.auth_sock`, but I think the environment variable approach is more general. besides, the implementation for a `auth_sock` config would have to set the environment variables anyway: as far as I know, there's no command line or config file options (other than the the `SSH_AUTH_SOCK` envvar) to specify an agent socket for the ssh clients.

**Additional context**

currently I am using a workaround by manually add a launcher menu entry, which in turn calls wezterm to connect to the ssh server:
```lua
config.launch_menu = {
	{
		label = "connect to server-a",
		args = { "wezterm", "connect", "server-a", },
		set_environment_variables = { SSH_AUTH_SOCK = "Z:/.ssh-agent.wsl1.sock", },
	},
}
config.ssh_domains = {
	{
		name = "server-a",
		remote_address = '12.34.56.78',
		username = "username",
	},
}
```
with this method, when I connect to a remote machine, there is always a new wezterm-gui window open, which may or may not be desired, but what annoys me most is that, for many times, I accidentally clicked the "attach domain server-a - ssh mux user@address" menu , and failed to connect, before I realized I should instead click the "connect to server-a" entry.

I should not have needed to use duplicated entries like these, if I can set the `SSH_AUTH_SOCK` variable in the ssh domain in the first place.

**Comment by wez** [comment]
It's worth noting that changing environment variables in a multi-threaded program has unpredictable effects, especially if multiple threads need to read that value around the time that you change it. For this reason, wezterm restricts any direct modification of its own-process environment to when it starts up, before it spawns any threads.  All other environment modification is done in terms of copying the current environment and then modifying the copy in preparation for launching child processes with that environment.

So, wezterm will not support updating any environment variable inside its own process after it has started up.

That said, I wonder if you are aware of the ssh `IdentityAgent` ssh_config option?

wezterm should read the `IdentityAgent` config option from your `~/.ssh/config` file, which is probably the best, most central place for you to configure this per destination host.

As mentioned in [the SshDomain docs](https://wezfurlong.org/wezterm/config/lua/SshDomain.html) you can specify ssh config overrides as part of the domain definition; those work for any supported config options, so you should be able to do something like this:

```lua
config.ssh_domains = {
  {
      name = "server-a",
      remote_address = '12.34.56.78',
      username = "username",
      ssh_option = {
         identityagent = 'Z:/.ssh-agent.wsl1.sock',
      },
  },
}
```

**Comment by wez** [comment]
Let's close it!

**Comment by nerditation** [comment]
well, actually I found another issue related to the handling of agent socket on Windows, it has something to do with either opening the socket with improper access rights (e.g. opened in exclusive mode or alike), or not properly shuting down the socket.

here's the situation: after I connect to a remote session via ssh, the agent socket won't accept new connection until the **wezterm-gui window quits**, note: simply ending the session (by detaching or exit all tabs of the session) is not enough, as long the same gui window is still open. I can't even re-connect to the just detached session if I don't quit the entire gui window.

this would also result lock situation if I try to connect to multiple remote sessions from the same window, then the second session would block on the auth socket, and the tab of the first session would hang on close, but weirdly the window would complete freeze if I try to detach the first session.

I tried the `wezterm ssh user@host` command, it would also block the auth socket.

and if I quit the agent, all the locked process would unlock (with failed authentication of course).

I am not complete sure it's an issue wezterm, but I think the agent is propably doing ok, while the AF_UNIX socket is blocked (meaning I can't use it in wsl1 either), the other protocols (plink and named pipe) work just fine. on the other hand, if I wasn't using wezterm, the wsl1 ssh would make as many connections with no problem, and, **wezterm can make its first connection even if there are active wsl1 ssh connections**, but as soon as wezterm connecs to the agent socket, on one else could use it until wezterm's gui window closed.
